File: ios/RnWmFetchGrpc/GrpcModule.swift
import Foundation
import GRPC
import NIO

@objc(GrpcModule)
class GrpcModule: NSObject {
  private var channel: GRPCChannel?
  private var group: EventLoopGroup?
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  @objc
  func initialize(_ host: String, port: Int, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    do {
      group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
      
      channel = try GRPCChannelPool.with(
        target: .host(host, port: port),
        transportSecurity: .plaintext,
        eventLoopGroup: group!
      )
      
      resolver(true)
    } catch {
      rejecter("INIT_ERROR", "Failed to initialize gRPC: \(error.localizedDescription)", error)
    }
  }
  
  @objc
  func createProduct(_ skuid: String, barcodePos: String, productName: String, merchantId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let channel = channel else {
      rejecter("NOT_INITIALIZED", "gRPC client not initialized", nil)
      return
    }
    
    // สร้าง gRPC client และเรียก RPC
    // let client = Product_ProductServiceClient(channel: channel)
    // var request = Product_CreateProductRequest()
    // request.skuid = skuid
    // request.barcodePos = barcodePos
    // request.productName = productName
    // request.merchantID = merchantId
    
    // let call = client.createProduct(request)
    
    // call.response.whenComplete { result in
    //   switch result {
    //   case .success(let response):
    //     let resultDict: [String: Any] = [
    //       "success": response.success,
    //       "message": response.message,
    //       "skuid": response.skuid
    //     ]
    //     resolver(resultDict)
    //   case .failure(let error):
    //     rejecter("GRPC_ERROR", "gRPC call failed: \(error.localizedDescription)", error)
    //   }
    // }
    
    // Placeholder for now
    let resultDict: [String: Any] = [
      "success": true,
      "message": "Product created (iOS stub)",
      "skuid": skuid
    ]
    resolver(resultDict)
  }
  
  @objc
  func shutdown(_ resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    channel?.close().whenComplete { _ in
      self.channel = nil
    }
    
    do {
      try group?.syncShutdownGracefully()
      group = nil
      resolver(true)
    } catch {
      rejecter("SHUTDOWN_ERROR", "Failed to shutdown: \(error.localizedDescription)", error)
    }
  }
}

File: android/app/src/main/java/com/rnwmfetchgrpc/GrpcModule.kt
package com.rnwmfetchgrpc

import com.facebook.react.bridge.*
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import com.rnwmfetchgrpc.grpc.ProductServiceGrpc
import com.rnwmfetchgrpc.grpc.CreateProductRequest
import com.rnwmfetchgrpc.grpc.CreateProductResponse

class GrpcModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    
    private val scope = CoroutineScope(Dispatchers.IO)
    private var channel: ManagedChannel? = null
    
    override fun getName(): String = "GrpcModule"
    
    @ReactMethod
    fun initialize(host: String, port: Int, promise: Promise) {
        try {
            channel = ManagedChannelBuilder
                .forAddress(host, port)
                .usePlaintext() // Use TLS in production
                .build()
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("INIT_ERROR", "Failed to initialize gRPC: ${e.message}", e)
        }
    }
    
    @ReactMethod
    fun createProduct(
        skuid: String?,
        barcodePos: String?,
        productName: String?,
        merchantId: String?,
        promise: Promise
    ) {
        scope.launch {
            try {
                val currentChannel = channel
                if (currentChannel == null) {
                    withContext(Dispatchers.Main) {
                        promise.reject("NOT_INITIALIZED", "gRPC client not initialized")
                    }
                    return@launch
                }
                
                // สร้าง gRPC stub และเรียก RPC
                val stub = ProductServiceGrpc.newBlockingStub(currentChannel)
                val request = CreateProductRequest.newBuilder()
                    .setSkuid(skuid ?: "")
                    .setBarcodePos(barcodePos ?: "")
                    .setProductName(productName ?: "")
                    .setMerchantId(merchantId ?: "")
                    .build()
                    
                val response: CreateProductResponse = stub.createProduct(request)
                
                withContext(Dispatchers.Main) {
                    val result = Arguments.createMap().apply {
                        putBoolean("success", response.success)
                        putString("message", response.message)
                        putString("skuid", response.skuid)
                    }
                    promise.resolve(result)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    promise.reject("GRPC_ERROR", "gRPC call failed: ${e.message}", e)
                }
            }
        }
    }
    
    @ReactMethod
    fun shutdown(promise: Promise) {
        try {
            channel?.shutdown()
            channel = null
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("SHUTDOWN_ERROR", "Failed to shutdown: ${e.message}", e)
        }
    }
}
